File: RUBY/DataStructures/graph.rb
require 'set'

class GraphNode
    attr_reader :value
    attr_accessor :neighbors
    def initialize(val)
        @value = val
        @neighbors = []
    end
end

def bfs(starting_node, target_value)
    visited = Set.new()
    queue = [starting_node]

    while queue.length > 0
        until !visited.include?(queue[0].value)
            queue.shift
            return nil if queue.length == 0
        end
        
        first = queue.shift

        return first if first.value == target_value
        
        queue += first.neighbors if first.neighbors
        visited.add(first.value)
    end
    return nil
end

graph = {
  'h': ['i', 'j'],
  'i': [],
  'j': ['k'],
  'k': [],
  'l': ['m'],
  'm': []
}

def dfs(graph)
    visited = Set.new();
    graph.keys.each do |key|
        _dfs(key, graph, visited)
    end
end

def _dfs(node, graph, visited)
    return nil if visited.include?(node.to_sym)

    puts node
    visited.add(node.to_sym)

    graph[node.to_sym].each do |neighbor|
        _dfs(neighbor, graph, visited)
    end
end

dfs(graph)

a = GraphNode.new('a')
b = GraphNode.new('b')
c = GraphNode.new('c')
d = GraphNode.new('d')
e = GraphNode.new('e')
f = GraphNode.new('f')
a.neighbors = [b,c,e]
c.neighbors = [b,d]
e.neighbors = [a]
f.neighbors = [e]

# p bfs(a, 'f')

File: JavaScript/DataStructures/graph.js
class GraphNode {
    constructor(val) {
        this.value = val;
        this.neighbors = [];
    }
}

function bfs(starting_node, target_value) {
    let visited = new Set();
    let queue = [starting_node];

    while(queue.length > 0) {
        while(visited.has(queue[0].value)){
            queue.shift();
            if(queue.length === 0) { return null }
        }

        let first = queue.shift();
        if(first.value === target_value) { return first }

        queue = queue.concat(first.neighbors);
        visited.add(first.value)
    }
    return null
}

const graph = {
    'a': ['b', 'c', 'e'],
    'b': [],
    'c': ['b', 'd'],
    'd': [],
    'e': ['a'],
    'f': ['e']
}

function dfs(node, graph, visited = new Set()) {
    if(visited.has(node)) { return null }

    console.log(node);
    visited.add(node);

    for(let i = 0; i < graph[node].length; i++) {
        dfs(graph[node][i], graph, visited)
    }
}

let a = new GraphNode('a');
let b = new GraphNode('b');
let c = new GraphNode('c');
let d = new GraphNode('d');
let e = new GraphNode('e');
let f = new GraphNode('f');

a.neighbors = [b,c,e];
// b.neighbors = [];
c.neighbors = [b,d];
// d.neighbors = [];
e.neighbors = [a];
f.neighbors = [e];

// console.log(bfs(a, "f"));
dfs('f', graph);
